stdlib/asm/crypto/fri.masm

```diff
@@ -0,0 +1,594 @@
+#! Note, both q ( polynomial evaluations i.e. in NTT form ), p ( polynomial in coefficient form ), is over quadratic 
+#! extension field over Fq | q = 2^64 - 2^32 + 1
+#!
+#! Other than starting address of q, starting memory address of p is also provided on stack, which is used for
+#! piping interpolated polynomial coefficients into memory, which will be later used for computing α.
+#!
+#! See https://github.com/0xPolygonMiden/miden-vm/issues/592#issuecomment-1341814867 for how τ is computed.
+#!
+#! Expected stack state
+#!
+#! [mem_ptr_q, mem_ptr_p, ...]
+#!
+#! Final stack state
+#!
+#! [τ1, τ0, mem_ptr_q, ...] | τ = (τ0, τ1)
+#!
+#! τ = rescue_prime_optimized::hash_elements(q... ++ p...)[..2]
+#!
+#! - `hash_elements` routine works over vector of elements of Fq | q = 2^64 -2^32 + 1
+#! - ... -> spread operator
+#! - ++ -> concatenate two vectors
+#! - q... -> vector of length 128 s.t. each element ∈ Fq
+#! - p... -> vector of length 16 s.t. each element ∈ Fq
+#! - Rescue Prime Optimized hash produces four Fq elements as its digest
+#! - [..2] -> picks first two Fq elements from digest
+proc.compute_tau
+    push.64
+    push.8
+    adv.ext2intt # interpolate polynomial p, using advice provider
+
+    drop
+    drop
+    dup
+
+    # Initialize Rescue Prime Optimized state
+    # More on RPO256 @ https://github.com/0xPolygonMiden/crypto/blob/3c60484/src/hash/rpo/mod.rs
+    push.0.0.0.0.0.0.0.0.0.0.0.0
+
+    # Consume `q` into Rescue state
+    repeat.16
+        mem_stream
+    end
+
+    movup.12
+    drop
+    movup.13
+    movdn.12
+
+    # Consume `p` into Rescue state
+    repeat.2
+        adv_pipe
+    end
+
+    movup.12
+    drop
+
+    dropw
+    swapw
+    dropw
+    drop
+    drop
+end
+
+#! Given ω, accumulator ν and τ on stack top, this routine first computes ⍳ ( using following formula ), 
+#! then it loads four elements from memory address (mem_ptr_q + 1) i.e. a3, a2, a1, a0 and finally accumulates
+#! it into ν, while consuming <a1, a0>, only when j is even | j ∈ [0..64), following formula for computing β,
+#! described in https://github.com/0xPolygonMiden/miden-vm/pull/644#pullrequestreview-1265141094
+#!
+#! Expected stack state
+#!
+#! [ω, ν1, ν0, τ1, τ0, mem_ptr_q - 1, ...]
+#!
+#! One computes ⍳ = ω/ (τ - ω)
+#! <a3, a2, a1, a0> loaded from (mem_ptr_q + 1) s.t. a1, a0 are consumed into ν
+#!
+#! Final stack state
+#!
+#! [a3, a2, ν1', ν0', τ1, τ0, mem_ptr_q, ...]
+#!
+#! Remaining part of word i.e. a3, a2 will be consumed during immediate next iteration
+#! of computing β which involves invocation of accumulate_for_odd_index.
+proc.accumulate_for_even_index
+    dup.4
+    dup.1
+    sub
+    dup.4
+    ext2inv
+    swap
+    dup.2
+    mul
+    movdn.2
+    mul
+
+    movup.6
+    add.1
+    movdn.6
+    push.0.0.0
+    dup.9
+    push.0
+    swap
+    mem_loadw
+
+    movup.5
+    movup.5
+    movup.5
+    movup.5
+    ext2mul
+    swap
+    movup.5
+    add
+    swap
+    movup.4
+    add
+    movdn.3
+    movdn.3
```

Comment:
I would add a short comment in front of each code block here to explain what it does. And similar comment for `accumulate_for_odd_index` below.

Reply:
Added in f0b1d17672364d01a1d69e210613ee5a85d2a8cb